test/e2e/util.go

```diff
@@ -527,49 +544,50 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
 	last = 0
 	failCount = 10
 	current = 0
+	podLists.Reset()
 	for same < failCount && current < replicas {
-		current = 0
-		waiting := 0
-		pending := 0
-		unknown := 0
-		time.Sleep(10 * time.Second)
-
-		currentPods, err := listPods(c, ns, label, fields.Everything())
-		if err != nil {
-			return fmt.Errorf("Error listing pods: %v", err)
-		}
-		if len(currentPods.Items) != len(pods.Items) {
-			return fmt.Errorf("Number of reported pods changed: %d vs %d", len(currentPods.Items), len(pods.Items))
-		}
-		for _, p := range currentPods.Items {
-			if p.Status.Phase == api.PodRunning {
-				current++
-				if err := VerifyContainersAreNotFailed(p, maxContainerFailures); err != nil {
-					return err
-				}
-			} else if p.Status.Phase == api.PodPending {
-				if p.Spec.Host == "" {
-					waiting++
-				} else {
-					pending++
+		time.Sleep(2 * time.Second)
+		for podLists.Len() > 0 && current < replicas {
+			item := podLists.Pop()
+			current = 0
+			waiting := 0
+			pending := 0
+			unknown := 0
+			currentPods := item.value.([]api.Pod)
+			for _, p := range currentPods {
+				if p.Status.Phase == api.PodRunning {
+					current++
+					if err := VerifyContainersAreNotFailed(p, maxContainerFailures); err != nil {
+						return err
+					}
+				} else if p.Status.Phase == api.PodPending {
+					if p.Spec.Host == "" {
+						waiting++
+					} else {
+						pending++
+					}
+				} else if p.Status.Phase == api.PodUnknown {
+					unknown++
 				}
-			} else if p.Status.Phase == api.PodUnknown {
-				unknown++
 			}
+			Logf("Pod States: %d running, %d pending, %d waiting, %d unknown", current, pending, waiting, unknown)
+			if podStatusFile != nil {
+				fmt.Fprintf(podStatusFile, "%s, %d, running, %d, pending, %d, waiting, %d, unknown\n", item.createTime, current, pending, waiting, unknown)
+			}
+			if last < current {
+				same = 0
+			} else if last == current {
+				same++
+			} else if current < last {
+				return fmt.Errorf("Number of running pods dropped from %d to %d", last, current)
+			}
+			if same >= failCount {
+				return fmt.Errorf("No pods started for the last %d checks", failCount)
+			}
+			last = current
 		}
-		Logf("Pod States: %d running, %d pending, %d waiting, %d unknown ", current, pending, waiting, unknown)
-		if last < current {
-			same = 0
-		} else if last == current {
-			same++
-		} else if current < last {
-			return fmt.Errorf("Number of running pods dropped from %d to %d", last, current)
-		}
-		if same >= failCount {
-			return fmt.Errorf("No pods started for the last %d checks", failCount)
-		}
-		last = current
 	}
+	close(stop)
```

Comment:
I feel like you have a bunch of return cases that will leak the lister goroutine, this diff is a little confusing and with all the comments it's hard to tell